src/twisted/words/protocols/irc.py

```diff
@@ -2620,9 +2622,17 @@ def connectionMade(self):
             self.register(self.nickname)
 
     def dataReceived(self, data):
-        basic.LineReceiver.dataReceived(self, data.replace('\r', ''))
+        data = data.replace('\r', '')
+        if isinstance(data, unicode):
+            data = data.encode("utf-8")
```

Comment:
IRC is a binary protocol without a defined encoding.  That means on Pythons > 2 this can produce incorrect behavior, because messages can be in any encoding, or even multiple encodings.  Users of this API wouldn't be able to do much about that.

The protocol should not be implemented in terms of native strings any more than HTTP should be parsed in terms of native strings.  In fact, it'd be better to do this with HTTP, which *does* have a defined encoding ([ISO 8859-1 for RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec2.html#sec2.2) and [ASCII for RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.2.4)), but Twisted does not.  It shouldn't do so here.

If this *does* do encoding and decoding, all the line length checks (e.g., [`_safeMaximumLineLength`](https://github.com/twisted/twisted/pull/534/files#diff-71c9155319411063da926db195c663d5R1684)) have to run after encoding, not before.  Otherwise a line of multi-byte UTF-8 code points will appear to be of an acceptable length but will actually be too long on the wire.